### ISSUE TITLE
AttachableQueryPart.getSQL(...) converts Float parameter to unexpected value

### ISSUE THREAD

**Issue by tomirio619** [jOOQ/jOOQ#17803]
**TLDR**: `AttachableQueryPart.getSQL(...)` converts float to unexpected value (in scientific notation)

Reproduction repository:  https://gitlab.com/sandsoft/reproduction-repositories/jooq-float-query-render

Assumed table definition:

```sql
CREATE TABLE review
(
    id     UUID,
    rating real
);
```

Given the record:

```text
+------------------------------------+------+
|id                                  |rating|
+------------------------------------+------+
|7af0c669-4802-4d40-91e9-24a4669f7236|0.03  |
+------------------------------------+------+
```

I want to build a query using jOOQ that retrieves all rows with a rating **greater than** `0.03`.
(I'm using a different library (JdbcTemplate) to execute the query).

The query is obtained as follows:

```kotlin
val query = selectFrom(REVIEW)
    .where(REVIEW.RATING.gt(savedReview.rating))
    .getSQL(ParamType.INLINED)
```

The result is:

```sql
SELECT
	public.review.id, public.review.rating
FROM
	public.review
WHERE
	public.review.rating > 2.9999999E-2;
```

This query incorrectly includes the record that I want to skip over (the one with `rating=0.03`).

If I execute the query with jOOQ, the generated SQL (as shown by jOOQ) also shows the value `2.9999999E-2`
However, the final (prepared statement) query executed against the database does have the correct value:

```sql
SELECT
	public.review.id, public.review.rating
FROM
	public.review
WHERE
	public.review.rating > $1;
-- parameters: $1 = '0.03'
```

### jOOQ Version

3.19.15

### Database product and version

PostgreSQL 15.7 on aarch64-unknown-linux-musl, compiled by gcc (Alpine 13.2.1_git20240309) 13.2.1 20240309, 64-bit

### Java Version

21

### JDBC / R2DBC driver name and version (include name if unofficial driver)

org.postgresql:postgresql:42.7.4

**Comment by lukaseder** [comment]
Ah yes, the wonders of IEEE 754 floating point arithmetic. But that is a correct value for `0.3f`. Try this:

```
jshell> 2.9999999E-2f == 0.03f
$1 ==> true
```

If you want exact decimal numbers, use `numeric`, not `real` (nor `double precision`).

The reason why this is rendered the way it is, is because we're using `DecimalFormat` to format floats:

```
jshell> import java.text.*

jshell> new DecimalFormat("0.#######E0", DecimalFormatSymbols.getInstance(Locale.US)).format(0.03f)
$3 ==> "2.9999999E-2"
```

**Comment by lukaseder** [comment]
To verify this within PostgreSQL, try this:

```sql
select 
  0.03::float4::text, 
  2.9999999E-2::float4::text,
  0.03::float4::text = 2.9999999E-2::float4::text
```

You'll get:

```
|text|text|?column?|
|----|----|--------|
|0.03|0.03|true    |
```

**Comment by tomirio619** [comment]
Thanks for your answer.

If I change the jOOQ query to explicitly cast the float value to a real, it indeed correctly skips over the record.

```kotlin
import org.jooq.impl.SQLDataType.REAL

val query = selectFrom(REVIEW)
    .where(REVIEW.RATING.gt(cast(savedReview.rating, REAL)))
    .getSQL(ParamType.INLINED)
```

Yields:

```sql
SELECT
	public.review.id, public.review.rating
FROM
	public.review
WHERE
	public.review.rating > CAST(2.9999999E-2 AS REAL);
```

Thanks for your help.

**Comment by lukaseder** [comment]
I see now, it seems that without the cast, the `rating` column gets promoted to `numeric`, which leads to the wrong result when using inline values. That's definitely a bug, then - we should automatically cast the value to `real`.

**Comment by lukaseder** [comment]
Ah yes, the assumption with rendering scientific notation was (among other things) that this effectively turns the value into a `float` (or `double`) literal. But that's not necessarily the case in some dialects, where `numeric` or `decimal` types also implement `decfloat` semantics.

**Comment by lukaseder** [comment]
Other, affected dialects include, at least (I'll amend the list):

- [x] Db2
- [x] Firebird
- [x] YugabyteDB

Affected, but cannot be fixed yet:

- MariaDB (see also: https://github.com/jOOQ/jOOQ/issues/10879)
- MySQL (see also: https://github.com/jOOQ/jOOQ/issues/10879)

Not affected, because the issue was fixed already in a different context, related to syntax:

- CockroachDB
- H2

**Comment by tomirio619** [comment]
Is it also fixed for PostgreSQL?

**Comment by lukaseder** [comment]
Yes of course

**Comment by lukaseder** [comment]
The generic fix leads to a regression in at least Db2:

```
org.jooq.exception.DataAccessException: SQL [create table "t" ("dummy" integer not null, "d" double default cast(0E0 as double), primary key ("dummy"))]; DB2 SQL Error: SQLCODE=-104, SQLSTATE=42601, SQLERRMC=as double;ble default cast(0E0;<space>, DRIVER=4.31.10
	at org.jooq_3.19.17-SNAPSHOT.DB2.debug(Unknown Source)
	at org.jooq.impl.Tools.translate(Tools.java:3620)
	at org.jooq.impl.Tools.translate(Tools.java:3608)
	at org.jooq.impl.DefaultExecuteContext.sqlException(DefaultExecuteContext.java:827)
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:362)
	at org.jooq.test.all.testcases.MetaDataTests.testMetaDataColumnDefault(MetaDataTests.java:1053)
	at org.jooq.test.jOOQAbstractTest.testMetaDataColumnDefault(jOOQAbstractTest.java:5259)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.testcontainers.containers.FailureDetectingExternalResource$1.evaluate(FailureDetectingExternalResource.java:29)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:316)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:240)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:214)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:155)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:385)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:507)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:495)
Caused by: com.ibm.db2.jcc.am.SqlSyntaxErrorException: DB2 SQL Error: SQLCODE=-104, SQLSTATE=42601, SQLERRMC=as double;ble default cast(0E0;<space>, DRIVER=4.31.10
	at com.ibm.db2.jcc.am.b7.a(b7.java:810)
	at com.ibm.db2.jcc.am.b7.a(b7.java:66)
	at com.ibm.db2.jcc.am.b7.a(b7.java:140)
	at com.ibm.db2.jcc.am.k_.c(k_.java:2844)
	at com.ibm.db2.jcc.am.k_.d(k_.java:2828)
	at com.ibm.db2.jcc.am.k_.b(k_.java:2188)
	at com.ibm.db2.jcc.t4.ab.k(ab.java:444)
	at com.ibm.db2.jcc.t4.ab.c(ab.java:102)
	at com.ibm.db2.jcc.t4.p.b(p.java:38)
	at com.ibm.db2.jcc.t4.av.h(av.java:124)
	at com.ibm.db2.jcc.am.k_.ak(k_.java:2183)
	at com.ibm.db2.jcc.am.k_.a(k_.java:3387)
	at com.ibm.db2.jcc.am.k_.e(k_.java:1135)
	at com.ibm.db2.jcc.am.k_.execute(k_.java:1114)
	at org.jooq.tools.jdbc.DefaultStatement.execute(DefaultStatement.java:102)
	at org.jooq.impl.SettingsEnabledPreparedStatement.execute(SettingsEnabledPreparedStatement.java:227)
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:458)
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:348)
	... 38 more
```
